src/werkzeug/formparser.py

```diff
@@ -423,156 +411,322 @@ def validate_boundary(self, boundary):
-        ``('begin_form', (headers, name))``
-        ``('begin_file', (headers, name, filename))``
-        ``('cont', bytestring)``
-        ``('end', None)``
+    class LineSplitter(object):
 
-        Always obeys the grammar
-        parts = ( begin_form cont* end |
-                  begin_file cont* end )*
-        """
-        next_part = b"--" + boundary
-        last_part = next_part + b"--"
-
-        iterator = chain(
-            make_line_iter(
-                file,
-                limit=content_length,
-                buffer_size=self.buffer_size,
-                cap_at_buffer=cap_at_buffer,
-            ),
-            _empty_string_iter,
-        )
+        """A stateful line splitter: call ``feed`` to push data in and receive
+        it split into lines.
 
-        terminator = self._find_terminator(iterator)
+        :param cap: Optional maximum length of a line; if this is given,
+                    lines will be truncated to meet it.
 
-        if terminator == last_part:
-            return
-        elif terminator != next_part:
-            self.fail("Expected boundary at start of multipart data")
-
-        while terminator != last_part:
-            headers = parse_multipart_headers(iterator)
+        ..versionadded:: 0.15
+        """
 
-            disposition = headers.get("content-disposition")
+        def __init__(self, cap=None):
+            self.buffer = b""
+            self.cap = cap
+
+        def _splitlines(self, pre, post):
+            # for most purposes, there is no difference between what we've
+            # just accepted and any other data. we accept them separately
+            # so we can tell if we're ending; see below.
+            buf = pre + post
+            rv = []
+            if not buf:
+                return rv, b""
+            lines = buf.splitlines(True)
+            iv = b""
+            for line in lines:
+                iv += line
+                while self.cap and len(iv) >= self.cap:
+                    rv.append(iv[: self.cap])
+                    iv = iv[self.cap :]
+                if line[-1:] in b"\r\n":
+                    rv.append(iv)
+                    iv = b""
+            # If this isn't the very end of the stream and what we got ends
+            # with \r, we need to hold on to it in case an \n comes next
+            if post and rv and not iv and rv[-1][-1:] == b"\r":
+                iv = rv[-1]
+                del rv[-1]
+            return rv, iv
+
+        def feed(self, data):
+            """Accepts a block of data to be split into lines, returning a list
+            of complete lines, including their terminator characters.
+
+            Feeding an empty block of data will end the parse, returning
+            an unterminated line, if any, and an empty string."""
+            lines, self.buffer = self._splitlines(self.buffer, data)
+            if not data:
+                lines += [self.buffer]
+                if self.buffer:
```

Comment:
A couple of odd things here. 

Since we want to reason here about `data` separately, and since `_splitlines` just concatenates `self.buffer` and `data`, it's more legible to do the concatenation here.

Second, it will be more convenient if `feed` is a generator, both for the consumer (which eventually needs to handle streaming data) and in terms of legibility here.

```
def feed(self, data):
  """docstring"""
  self.buffer += data
  lines, self.buffer = self._splitlines(self.buffer)
  for line in lines:
    yield line
  if not data:
    yield self.buffer
  if self.buffer:
    yield b""
```

(note that `_splitlines` should only accept one argument)